MdeModulePkg/Library/DxeCorePerformanceLib/DxeCorePerformanceLib.c

```diff
@@ -307,26 +314,39 @@ InternalGetSmmPerfData (
           SmmBootRecordData             = AllocateZeroPool (SmmBootRecordDataSize);
           SmmBootRecordDataRetrieved    = 0;
           ASSERT (SmmBootRecordData  != NULL);
-          SmmCommData->BootRecordData = (VOID *)((UINTN)SmmCommMemRegion->PhysicalStart + SMM_BOOT_RECORD_COMM_SIZE);
-          SmmCommData->BootRecordSize = ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE;
-          while (SmmBootRecordDataRetrieved < SmmBootRecordDataSize) {
-            Status = Communication->Communicate (Communication, SmmBootRecordCommBuffer, &CommSize);
-            ASSERT_EFI_ERROR (Status);
-            ASSERT_EFI_ERROR (SmmCommData->ReturnStatus);
-            if (SmmBootRecordDataRetrieved + SmmCommData->BootRecordSize > SmmBootRecordDataSize) {
-              CopyMem ((UINT8 *)SmmBootRecordData + SmmBootRecordDataRetrieved, SmmCommData->BootRecordData, SmmBootRecordDataSize - SmmBootRecordDataRetrieved);
-            } else {
-              CopyMem ((UINT8 *)SmmBootRecordData + SmmBootRecordDataRetrieved, SmmCommData->BootRecordData, SmmCommData->BootRecordSize);
+          if (SmmBootRecordData != NULL) {
+            CurrentBootRecordData = (VOID *)((UINTN)SmmCommMemRegion->PhysicalStart + SMM_BOOT_RECORD_COMM_SIZE);
+            while (SmmBootRecordDataRetrieved < SmmBootRecordDataSize) {
+              // Note: Maximum comm buffer data payload size is ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE
+              BootRecordDataPayloadSize = MIN (
+                                            ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE,
+                                            SmmBootRecordDataSize - SmmBootRecordDataRetrieved
+                                            );
+
+              MmCommBufferHeader->MessageLength = sizeof (SMM_BOOT_RECORD_COMMUNICATE) + BootRecordDataPayloadSize;

```

Comment:
do you find the original code has the bug for the MmCommBufferHeader->MessageLength? if so, suggest use another commit for the bug fix.

Reply:
In the original code: `MmCommBufferHeader->MessageLength = sizeof (SMM_BOOT_RECORD_COMMUNICATE);` was valid because the comm buffer data was always a `SMM_BOOT_RECORD_COMMUNICATE` structure and the actual data was always in a buffer pointed to by the `BootRecordData` structure member.

We needed a model without nested pointers in the comm buffer. While the MM handler code (`FpdtSmiHandler()`) supports both approaches for backward compatibility, in the DXE Core, we use the new approach which sets `SmmCommData->BootRecordData = NULL` to indicate that the data should be added after the `SMM_BOOT_RECORD_COMMUNICATE` structure in the comm buffer.

Reply:
Going to close this thread unless you see a need to reopen.